src/Factory/AuthFactory.php

```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace Transloadit\Factory;
+
+use Transloadit\Model\Auth;
+use DateTimeInterface;
+use Transloadit\Model\Contracts\AuthInterface;
+
+class AuthFactory
+{
+    public static function create(string $key, string $secret, DateTimeInterface $expires = null): AuthInterface
+    {
+        return new Auth($key, $secret, $expires);
```

Comment:
I don't really understand the purpose of all the Factory classes. I feel like just having `new Auth($key, $secret)` in the code is more readable and writable than `AuthFactory::create($key, $secret)`. Is there some benefit that I am missing?

Reply:
Hi @goto-bus-stop first of all, thank you so much for your time to try to review my PR

regarding about your questions 

> from what I remember that is very useful with dependency injection in frameworks like Laravel.

Exactly ❤️ (`DIP` and `ISP`)

> IMO there should be a way to provide Assembly instructions as an array.
Well use array in this point I'm not sure if it's the best approach

1 - you need to validate the data that is inputed in your method, because you can receive any data.
2 - Always you need to check the structure to pass the right "key"

IMO is better to use `Object` instead of `array`

Other point is, using array is more complicated you set default parameters like  `templates`, `credentials` and so on ...

if you see [here](https://github.com/eerison/myprofile/blob/master/src/Service/CurriculumService.php#L34-L52), always I need to pass `template` but this parameter never change.

in my example you can use this like 

```php
$parameter = new Parameter([... steps], 'templateId');

//or 

$parameter->setTemplateId('1234')
```

this way you can pass your `resource` already configured in yours services, you just need to pass your steps like

just a example

```php
class YourService
{
    private $assembly;
    private $assemblyResourceService;
    
    public function __controller(AssemblyInterface $assembly, AssemblyResourceServiceInterface $assemblyResourceService)
    {
        $this->assembly = $assembly;
        $this->assemblyResourceService = $assemblyResourceService;
    }
    
    public function your_method()
    {
        $step1 = StepFactory::create('resize', [
            'robot' => '/image/resize',
            'width' => 200,
            'height' => 100,
        ]);
        
        $assembly = $this->assembly;
        // and here there are parameters like templateId, notifyUrl, allowStepsOverride and so on
        $assembly->getParameter()
            ->addStep($step1)
            ->addStep($step2)
            ///....
            ;
        $assembly->addFilePath('your_path.img');
            
       $this->assemblyResourceService->create($assembly);
    }
}
```

> I don't really understand the purpose of all the Factory classes. I feel like just having new Auth($key, $secret) in the code is more readable and writable than AuthFactory::create($key, $secret). Is there some benefit that I am missing?

I did it just to avoid you coupling code like `new Step();`

and the others factories like `AuthFactory`, `AssemblyFactory` etc ... I guess it will be used in standalone application only, not in applications that use symfony, laravel and so on.

Reply:
Thank you, that code snippets clears some things up for me about how the AssemblyInterface is intended to be used. (I was still thinking that you'd create them manually inside controllers or service classes rather than injecting them together with the `resource->create()` one.)

Your point about objects over arrays makes sense, but I also don't think we really need to validate the array input … the Transloadit API will reject invalid inputs, so people who still use the arrays would just run into an error later down the line. What I'd hope for is that we could recommend using the new object style but have the array style as a migration path from 3.x. (so users only have to update their function calls and not also rewrite the templates etc)

> I did it just to avoid you coupling code like new Step();

That does feel like unnecessary abstraction IMHO…or is that what all modern PHP libraries do?

Reply:
Hey @goto-bus-stop 

> Thank you, that code snippets clears some things up for me about how the AssemblyInterface is intended to be used. (I was still thinking that you'd create them manually inside controllers or service classes rather than injecting them together with the `resource->create()` one.)
> 
> Your point about objects over arrays makes sense, but I also don't think we really need to validate the array input … the Transloadit API will reject invalid inputs, so people who still use the arrays would just run into an error later down the line. What I'd hope for is that we could recommend using the new object style but have the array style as a migration path from 3.x. (so users only have to update their function calls and not also rewrite the templates etc)

I got it your point ....

I'll see some way to accept `object` and `array` to accept migrate the old version, and I'll post here :)

in the version `4.1.0` I will add Deprecation that will be removed in version `5`

> That does feel like unnecessary abstraction IMHO…or is that what all modern PHP libraries do?

I guess it's ok to remove those factories, because usually devs uses some framework to do `Dependency injection`